Automattic/jetpack
# Social: Enforce sharing limits in the block editor

This PR enforces sharing limits in the block editor, if applicable.

There's an edge case that makes it possible to bypass the limit by adding a new connection during the publishing process—we can fix this in a follow-up PR.

#### Changes proposed in this Pull Request:

* Check if the sharing limit is enabled for a specific site.
* If it is, ensure users cannot enable more connections than they have shares left.
* Disable connections by default in the block editor when the number of connections is greater than the number of shares left.

#### Other information:


#### Jetpack product discussion


pdrWKz-wQ-p2
p6TEKc-6td-p2

#### Does this pull request change what data or activity we track or use?


No.

#### Testing instructions:






_Please note that this PR needs D85993-code to be merged. In the meantime, you can sandbox it to test._

* Make sure tests pass: `jetpack test plugins/social php`
* Enable the Jetpack Social plugin.
* Make a new post, and check that the sharing screen looks like the first image.
* Enable sharing limits for your site (first line in test plan of D85993-code)
* Make a new post, and check that you can see the upgrade notice (second image).
* Test that connections are disabled by default when there's more connections than shares left, and that you can't select more connections than allowed.
  * To mock the number of shares left, you can add `$shares['shares_remaining'] = 1;` to `initial_state()` and `$shares_remaining = 1;` to `set_up_sharing_limits()` in `jetpack-social.php`
* Finally, enable the Jetpack plugin, and make sure the screen looks like the first screenshot again

<table>
<tr>
	<td>Sharing Limits Disabled 
	<td>Sharing Limits Enabled
	<td>More connections than shares left
	<td>More connections than shares left, and all remaining shares have been used
<tr>
	<td><img width="278" alt="image" src="https://user-images.githubusercontent.com/1713699/184976111-03be1d83-04b1-449c-bdea-583b8281a292.png">
	<td><img width="278" alt="image" src="https://user-images.githubusercontent.com/1713699/184976450-e0c32894-9eb3-4402-a5ea-826f367c2d25.png">
	<td><img width="280" alt="image" src="https://user-images.githubusercontent.com/1713699/184976463-949ade72-0653-4c15-95a3-e20fa62945a4.png">
	<td><img width="280" alt="image" src="https://user-images.githubusercontent.com/1713699/184976475-d15d5279-9561-4d0f-b536-f7b34e431524.png">
</table>
